### ISSUE TITLE
MPI_Comm_idup hang in v2.0.x with MPI_THREAD_MULTIPLE

### ISSUE THREAD

**Issue by jjhursey** [open-mpi/ompi#2380]
I created a test case in ompi-tests under`ibm/communicator/comm_idup_many.c`

This test is doing a simple:
```c
    for (i = 0; i < num_comms_to_create; i++) {
        MPI_Comm_idup(MPI_COMM_WORLD, &comms[i], &req[i]);
    }

    MPI_Waitall(num_comms_to_create, req, MPI_STATUSES_IGNORE);
```

If you compile the `v2.0.x` branch with `--enable-mpi-thread-multiple` this test will hang when you try to create more than one communicator this way.
```shell
./configure --enable-debug --enable-mpi-thread-multiple
```

The problem is that the first request will progress past stage one, but the second request will block the progress thread in this section of code:
 * https://github.com/open-mpi/ompi/blob/v2.0.x/ompi/communicator/comm_cid.c#L517-L530

The `master` and `v2.x` branches pass this test. So I believe that the CID allocation fix that went into those two branches 'fixed' this problem.

**Comment by jjhursey** [comment]
As part of Issue #2234 it was requested that we port a portion of this change to the `v2.0.x` series to fix a spawn issue:
 * https://github.com/open-mpi/ompi/issues/2234#issuecomment-257398856

I believe that this commit to the `v2.x` branch is what we would be looking at:
 * https://github.com/open-mpi/ompi/pull/2215

**Comment by jsquyres** [comment]
Per the call today, https://github.com/open-mpi/ompi/pull/2381 may fix this issue.

**Comment by jsquyres** [comment]
Confirmed by @jjhursey: fixed by #2381.
